tools/setup_services.py

```diff
@@ -1,15 +1,396 @@
 from baselayer.log import make_log
 
 log = make_log("baselayer")
 
 
-def copy_supervisor_configs():
-    env, cfg = load_env()
+def generate_supervisor_config(service_name, service_path):
+    """
+    Generates a supervisor configuration for a given service.
+    """
+    supervisor_conf_template = f"""
+[program:{service_name}]
+command=/usr/bin/env python {pjoin(service_path, 'main.py')} %(ENV_FLAGS)s
+environment=PYTHONPATH=".",PYTHONUNBUFFERED="1"
+stdout_logfile=log/{service_name}_service.log
+redirect_stderr=true
+"""
+    # Write the configuration to a file
+    with open(f"{service_path}/supervisor.conf", "w") as f:
+        f.write(supervisor_conf_template)
+
+
+def read_plugin_config(plugin_path):
+    # read the config (pyproject.toml) for the plugin
+    config_path = pjoin(plugin_path, "pyproject.toml")
+    with open(config_path, "rb") as f:
+        return tomli.load(f)
+
+
+def get_plugin_compatible_version(plugin_config: dict):
+    # if the plugin specifies a "tool.<name>" section with
+    # a version requirement, retrive it
+    if "tool" not in plugin_config:
+        log("Plugin config does not contain 'tool' section")
+        return None, None
+    # there should only be one key in tool, which is the name of the package
+    # for which we are installing the plugin
+    if len(plugin_config["tool"]) != 1:
+        log("Plugin config must have exactly one 'tool' section")
+        return None, None
+
+    tool_name = next(iter(plugin_config["tool"]))
+    tool_version_requirement = plugin_config["tool"][tool_name].get("version", None)
+
+    if not tool_version_requirement:
+        log(f"Plugin config for '{tool_name}' does not specify a version")
+        return None, None
+
+    return tool_name, tool_version_requirement
+
+
+def validate_version(version_string, requirement_string):
+    """
+    Validate a version against a version requirement.
+
+    Args:
+        version_string (str): The version to check (e.g., "1.2.3")
+        requirement_string (str): The requirement specification (e.g., ">=1.0.0", "==1.2.3", ">=1.0,<2.0")
+
+    Returns:
+        bool: True if version satisfies the requirement, False otherwise
+    """
+    try:
+        v = version.parse(version_string)
+        spec = SpecifierSet(requirement_string)
+        return v in spec
+    except Exception as e:
+        log(f"Error validating version: {e}")
+        return False
+
+
+def validate_plugin_compatibility(plugin_name: str, plugin_path: str):
+    plugin_config = read_plugin_config(plugin_path)
+    name, version_requirement = get_plugin_compatible_version(plugin_config)
+    if not (name and version_requirement):
+        return True
+    try:
+        mod = __import__(name)
+        installed_version = mod.__version__
+
+        if not validate_version(installed_version, version_requirement):
+            log(
+                f"Plugin {plugin_name} is incompatible: required {version_requirement}, found {installed_version}. Skipping."
```

Comment:
"incompatible" is a bit ambiguous here; what is incompatible with what.

Would be helpful if we could say something like:

"Plugin myplugin requires skyportal x, but this is skyportal y. Skipping."

Reply:
Agreed, we'll clarify all the return messages.

Reply:
done